runtime/integration-tests/src/generic/utils/currency.rs

```diff
@@ -8,128 +8,158 @@ use cfg_types::{
-	const UNIT: Balance = 10u128.pow(Self::DECIMALS);
-	const SYMBOL: &'static str;
-	const NAME: &'static str = Self::SYMBOL;
-	const LOCATION: Option<xcm::VersionedMultiLocation> = None;
-	const CUSTOM: CustomMetadata;
-	const ED: Balance = 0;
-
-	fn metadata() -> AssetMetadata<Balance, CustomMetadata> {
+	fn id(&self) -> CurrencyId;
+
+	fn decimals(&self) -> u32 {
+		18
+	}
+
+	fn unit(&self) -> Balance {
+		10u128.pow(self.decimals())
+	}
+
+	fn symbol(&self) -> &'static str;
+
+	fn name(&self) -> &'static str {
+		&self.symbol()
+	}
+
+	fn location(&self) -> Option<xcm::VersionedMultiLocation> {
+		None
+	}
+
+	fn custom(&self) -> CustomMetadata;
+
+	fn ed(&self) -> Balance {
+		0
+	}
+
+	fn metadata(&self) -> AssetMetadata<Balance, CustomMetadata> {
 		AssetMetadata {
-			decimals: Self::DECIMALS,
-			name: Self::NAME.as_bytes().to_vec(),
-			symbol: Self::SYMBOL.as_bytes().to_vec(),
-			existential_deposit: Self::ED,
-			location: None,
-			additional: CustomMetadata {
-				pool_currency: true,
-				..Default::default()
-			},
+			decimals: self.decimals(),
+			name: self.name().as_bytes().to_vec(),
+			symbol: self.symbol().as_bytes().to_vec(),
+			existential_deposit: self.ed(),
+			location: self.location(),
+			additional: self.custom(),
 		}
 	}
 }
 
 /// Matches default() but for const support
-const CONST_DEFAULT_CUSTOM: CustomMetadata = CustomMetadata {
+pub const CONST_DEFAULT_CUSTOM: CustomMetadata = CustomMetadata {
 	transferability: CrossChainTransferability::None,
 	mintable: false,
 	permissioned: false,
 	pool_currency: false,
 	local_representation: None,
 };
 
-pub fn find_metadata(currency_id: CurrencyId) -> AssetMetadata<Balance, CustomMetadata> {
-	match currency_id {
-		Usd6::ID => Usd6::metadata(),
-		Usd12::ID => Usd12::metadata(),
-		Usd18::ID => Usd18::metadata(),
-		_ => panic!("Unsupported currency {currency_id:?}"),
-	}
-}
-
 pub fn price_to_currency<N: FixedPointNumber<Inner = Balance>>(
 	price: N,
-	currency_id: CurrencyId,
+	currency_id: impl CurrencyInfo,
 ) -> Balance {
-	let decimals = find_metadata(currency_id).decimals;
-	conversion::fixed_point_to_balance(price, decimals as usize).unwrap()
+	conversion::fixed_point_to_balance(price, currency_id.decimals() as usize).unwrap()
 }
 
 pub struct Usd6;
 impl CurrencyInfo for Usd6 {
-	const CUSTOM: CustomMetadata = CustomMetadata {
-		pool_currency: true,
-		transferability: CrossChainTransferability::Xcm(XcmMetadata {
-			fee_per_second: Some(1_000),
-		}),
-		..CONST_DEFAULT_CUSTOM
-	};
-	const DECIMALS: u32 = 6;
-	const ID: CurrencyId = CurrencyId::ForeignAsset(1);
-	const SYMBOL: &'static str = "USD6";
+	fn id(&self) -> CurrencyId {
+		CurrencyId::ForeignAsset(1)
+	}
+
+	fn decimals(&self) -> u32 {
+		6
+	}
+
+	fn symbol(&self) -> &'static str {
+		"USD6"
+	}
+
+	fn custom(&self) -> CustomMetadata {
+		CustomMetadata {
+			pool_currency: true,
+			transferability: CrossChainTransferability::Xcm(XcmMetadata {
+				fee_per_second: Some(1_000),
+			}),
```

Comment:
Just wondering why `Usd6` has this property but `Usd12` and `Usd18` do not. Could we use the same for all of them?

Reply:
I just copy pasted everything. Not sure, why it was there.

Reply:
Seems like we want to have default transferability here instead?